### PR TITLE
IMXRT1020_EVK: Add infrastructure to flash raw binary

### PR DESCRIPTION
**Pull request by btheosam**
New file: bootstrap.sh to bootstrap into flash loader
modified: imxrt1020-evk_download.sh
new file: ivt_flashloader.bin bootstrap binary
new file: blhost & sdphost (Linux tools for imxrt1020 board)

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [ea859748ded7328048b03bd8410ffc8d0b9d7619] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [ea859748ded7328048b03bd8410ffc8d0b9d7619] - Code Rule Check OK.

File: build/configs/imxrt1020-evk/bootstrap.sh [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+SDPHOST=${CURDIR_PATH}/sdphost
+BLHOST=${CURDIR_PATH}/blhost
+BINFILE=${CURDIR_PATH}/ivt_flashloader.bin
+
+BASEADDR=0x20208000
+IVTADDR=0x20208400
+TTYDEV="/dev/ttyACM0"
+SUDO=sudo
+
+echo "Connecting the chip ROM ... "
```
Comment (sijupunnnoose) [inline build/configs/imxrt1020-evk/bootstrap.sh]: Have some check for connected board and build config used, currently if build for 1050 config and try to flash on 1020 there is no proper error status.

Comment (btheosam) [reply build/configs/imxrt1020-evk/bootstrap.sh]: fixed

Comment (btheosam) [reply build/configs/imxrt1020-evk/bootstrap.sh]: fixed

**Comment by seinfra** [comment]
Target : [da6a0a76ca8d4e03ad31fd2de164489f36152b05] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [da6a0a76ca8d4e03ad31fd2de164489f36152b05] - Code Rule Check (C++) OK.

File: build/configs/imxrt1020-evk/bootstrap.sh (hunk 1 above)
Comment (sijupunnnoose) [reply build/configs/imxrt1020-evk/bootstrap.sh]: I'm not getting any error message, can you share the failure logs here?

Comment (sijupunnnoose) [reply build/configs/imxrt1020-evk/bootstrap.sh]: ########################################
 WARNING:If download fails, don't forget to replug
 and reset the board before next try!!
 #################################################

Using partitions from .config file

Using sizes from .config file

offsets: 0x60000000 0x60040000 0x60080000 0x600C0000 0x60100000 0x60140000
sizes: 256 256 256 256 256 256
partitions: kernel app micom micom wifi wifi
Connecting the chip ROM ...
Status (HAB mode) = 1450735702 (0x56787856) HAB disabled.
Reponse Status = 4042322160 (0xf0f0f0f0) HAB Success.

Download /home/siju.cherian/Tizen/work/TizenRT/TizenRT/build/configs/imxrt1050-evk/../../tools/nxp/ivt_flashloader.bin to 0x20000000
Preparing to send 90039 (0x15fb7) bytes to the target.
(1/1)1%unexpected write file response 0x89232389.
Status (HAB mode) = 1450735702 (0x56787856) HAB disabled.
Reponse Status = 2300781449 (0x89232389) REENUM Ack

Start at 0x20000400
Status (HAB mode) = 1450735702 (0x56787856) HAB disabled.

wait ...
Error: Initial ping failure: No response received for ping command.Makefile.unix:533: recipe for target 'download' failed
make: *** [download] Error 1

Comment (btheosam) [reply build/configs/imxrt1020-evk/bootstrap.sh]: bsvt18@ubuntu:~/TizenRT/PublicGit/June6/os$ make download

########################################
 WARNING:Make sure NO other application like 
 putty etc, is accessing ttyACM0. If download 
 fails, reset the board before next try!!
 #################################################

Target is NOT IMXRT1020_EVK
Makefile.unix:533: recipe for target 'download' failed
make: *** [download] Error 1

Comment (btheosam) [reply build/configs/imxrt1020-evk/bootstrap.sh]: The actual code that validates this is 
if [[ "${CONFIG_ARCH_BOARD_IMXRT1020_EVK}" != "y" ]]
FIND_WORD="RT1020-EVK"	then
	echo "Target is NOT IMXRT1020_EVK"
	exit 1
fi

File: build/configs/imxrt1020-evk/imxrt1020-evk_download.sh [hunk 2]
```diff
@@ -18,34 +18,167 @@
...
 
-IMXRT1020_USB_STRING=`mount | grep RT1020-EVK`
-FIND_WORD="RT1020-EVK"
+if [[ "${CONFIG_ARCH_BOARD_IMXRT1020_EVK}" != "y" ]]
+then
+	echo "Target is NOT IMXRT1020_EVK"
+	exit 1
+fi
+
+if [ ! -f ${TINYARA_BIN} ]; then
```
Comment (sunghan-chang) [inline build/configs/imxrt1020-evk/imxrt1020-evk_download.sh]: Where is `TINYARA_BIN` definition?

**Comment by seinfra** [comment]
Target : [795a766b41c48d3a0ceac115b2a1a6c223b7921c] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [795a766b41c48d3a0ceac115b2a1a6c223b7921c] - Code Rule Check OK.

File: build/configs/imxrt1020-evk/imxrt1020-evk_download.sh (hunk 2 above)
Comment (btheosam) [reply build/configs/imxrt1020-evk/imxrt1020-evk_download.sh]: Updated with definition. Thanks for the catch.

Comment (sunghan-chang) [reply build/configs/imxrt1020-evk/imxrt1020-evk_download.sh]: I think you have to get the path of tinyara bin.
```
if [ ! -f ${OUTBIN_PATH}/${TINYARA_BIN} ]; then

or

TINYARA_BIN="${OUTBIN_PATH}/tinyara.bin"
```

Comment (btheosam) [reply build/configs/imxrt1020-evk/imxrt1020-evk_download.sh]: Oh yes, have to be fixed and tested more.

Comment (btheosam) [reply build/configs/imxrt1020-evk/imxrt1020-evk_download.sh]: With latest fix:
########################################
 WARNING:Make sure NO other application like 
 putty etc, is accessing ttyACM0. If download 
 fails, reset the board before next try!!
 #################################################

missing file /home/bsvt18/TizenRT/PublicGit/June6/build/configs/imxrt1020-evk/../../../build/output/bin/tinyara.bin
Makefile.unix:533: recipe for target 'download' failed
make: *** [download] Error 1

**Comment by seinfra** [comment]
Target : [c297b263783f19f987ae9f1fef7e113e982bc81e] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [c297b263783f19f987ae9f1fef7e113e982bc81e] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [7e79567774df4dbc8cec26776e4854836bc13bb6] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [7e79567774df4dbc8cec26776e4854836bc13bb6] - Code Rule Check (C++) OK.

File: build/configs/imxrt1020-evk/imxrt1020-evk_download.sh [hunk 3]
```diff
@@ -18,34 +18,169 @@
...
 
-GET_FILENAME
-echo "Downloading ${FLASH_IMG}"
-cp ${FLASH_IMG} ${IMXRT1020_USB_PATH} || exit 1
+##Utility function to match partition name to binary name##
+
+#NOTE: Hard-coded for now
+function get_executable_name()
+{
+	case $1 in
```
Comment (btheosam) [inline build/configs/imxrt1020-evk/imxrt1020-evk_download.sh]: This should cover imxrt partition as well

Comment (btheosam) [reply build/configs/imxrt1020-evk/imxrt1020-evk_download.sh]: resolved as there is NO binary for imxrt now

Comment (btheosam) [reply build/configs/imxrt1020-evk/imxrt1020-evk_download.sh]: There isn't any binary for imxrt now. Hence resolved.

File: build/configs/imxrt1020-evk/bootstrap.sh [hunk 4]
```diff
@@ -0,0 +1,49 @@
+#!/bin/bash -e
+
+
+CURDIR=$(readlink -f "$0")
+CURDIR_PATH=$(dirname "$CURDIR")
+
+BINFILE=${CURDIR_PATH}/ivt_flashloader.bin
+SDPHOST=${CURDIR_PATH}/sdphost
+BLHOST=${CURDIR_PATH}/blhost
+BINFILE=${CURDIR_PATH}/ivt_flashloader.bin
```
Comment (kishore-sn) [inline build/configs/imxrt1020-evk/bootstrap.sh]: BINFILE definition is repeated here.

File: build/configs/imxrt1020-evk/imxrt1020-evk_download.sh (hunk 3 above)
Comment (sunghan-chang) [reply build/configs/imxrt1020-evk/imxrt1020-evk_download.sh]: We have to erase userfs region as well to support smartfs. SMARTFS does not erase that region before using.

Comment (sunghan-chang) [reply build/configs/imxrt1020-evk/imxrt1020-evk_download.sh]: And we had better rename the partition name of smartfs from imxrt to userfs in all of defconfigs.

Comment (btheosam) [reply build/configs/imxrt1020-evk/imxrt1020-evk_download.sh]: Okay.

**Comment by seinfra** [comment]
Target : [98d8c56986c6d42eca232d6514fc2ddf24e854c1] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [98d8c56986c6d42eca232d6514fc2ddf24e854c1] - Code Rule Check (C++) OK.

File: build/configs/imxrt1020-evk/bootstrap.sh (hunk 4 above)
Comment (btheosam) [reply build/configs/imxrt1020-evk/bootstrap.sh]: fixed

**Pull request merged**